ViewEnvironment/Sources/EnvironmentValues+ViewEnvironment.swift

```diff
@@ -29,4 +29,28 @@ extension EnvironmentValues {
     }
 }
 
+
+extension Environment where Value == ViewEnvironment {
+    
+    @available(
+        *,
+         deprecated,
+         message:
+            """
+            Please do not create an `@Environment` property that references the top-level `viewEnvironment`: \
+            it will break SwiftUI's automatic invalidation when any part of the `ViewEnvironment` changes. \
```

Comment:
for my own personal edification: can you elaborate on the risk & consequences a bit more? what is the behavioral difference in these 2 approaches?

Reply:
So basically, the way that SwiftUI does state invalidation is to compare the properties on a view between the old view and the new view, and if they're different or otherwise not comparable, the view is redrawn and updated.

Because `ViewEnvironment` is a big opaque bag o' crap, its possible that there's non-comparable stuff in there, or there's stuff in there that changes every render cycle, meaning that it'd lead to view invalidation. So the fix for that is to scope your `Environment` property down to just the data you need.

I _know_ there's a point free or objc.io article about this in more depth somewhere, however I am struggling to find it.

Reply:
thanks! this makes sense – i had failed to realize that the root `.viewEnvironment` key is like putting a copy of _another_ environment into the... `SwiftUI.Environment`, and clearly you don't want to depend on the entire environment itself, just the stuff your view 'cares about'.